README.md
# BxTextField

This component available put formatted text, for example: phone number, url or email address, credit card with comfortable format. Also you can use other features what we have in `UITextField` whitout issues from Apple.

## Gif demo

![Demo Gif](https://github.com/ByteriX/BxTextField/raw/master/Screenshots/BxTextFieldDemo.gif "Demo Gif")

## Features

- [x] Inherited from `UITextField` and don't use delegate
- [x] Can use patterns sides with constant text
- [x] Have formatting putting with different filling direction
- [x] Correct working with selecting/editing text
- [x] Fixed native issues (problems with edges)
- [x] Covered of Tests

## Requirements

- iOS 8.0+ : iOS 8.x/9.x/10.x/11.x/12.x/13.x/14.x
- Swift 3.0+ : Swift 3.2/4.x/5.x supported

## Installation

### CocoaPods

[CocoaPods](http://cocoapods.org) is a dependency manager for Cocoa projects. You can install it with the following command:

```bash
$ gem install cocoapods
```

To integrate BxTextField into your Xcode project using CocoaPods, specify it in your `Podfile`:

```ruby
source 'https://github.com/CocoaPods/Specs.git'
platform :ios, '8.0'
use_frameworks!

target '<Your Target Name>' do
pod 'BxTextField', '~> 1.4'
end
```

Then, run the following command:

```bash
$ pod install
```


### Swift Package Manager

The [Swift Package Manager](https://swift.org/package-manager/) is a tool for automating the distribution of Swift code and is integrated into the `swift` compiler. It is in early development, but BxTextField does support its use on supported platforms. 

Once you have your Swift package set up, adding BxTextField as a dependency is as easy as adding it to the `dependencies` value of your `Package.swift`.

```swift
dependencies: [
    .Package(url: "https://github.com/ByteriX/BxTextField.git", majorVersion: 1)
]
```

### Manually

If you prefer not to use either of the aforementioned dependency managers, you can integrate `BxTextField` into your project manually.

#### Embedded Framework

- Open up Terminal, `cd` into your top-level project directory, and run the following command "if" your project is not initialized as a git repository:

```bash
$ git init
```

- Add `BxTextField` as a git [submodule](http://git-scm.com/docs/git-submodule) by running the following command:

```bash
$ git submodule add https://github.com/ByteriX/BxTextField.git
```

- Add all sources and resources from local copy of `BxTextField` to the build phase of the project.

- And that's it!

## Properties

### Formatting

- formattingTemplate: String, Format text for putting pattern. If formattingReplacementChar is "*" then example may has value "**** - **** - ********". Default is ""

- formattingReplacementChar: String, Replacement symbol, it use for formattingTemplate as is as pattern for replacing. Default is "#"
- formattingEnteredCharacters: String, Allowable symbols for entering. Uses only if formattingTemplate is not empty. Default is "", that is all symbols.
- formattingDirection: FormattingDirection, Direction for replacement text from template. Can equal leftToRight or rightToLeft value. Default is leftToRight
- isFormattingRewriting: Bool, When user would try put extra symboles for filled text and if it's true then text will be rewrited. It depends from formattingDirection too.

### Patterns

- rightPatternText: String, Not editable pattern part of the text. Defaults to "".
- leftPatternText: String, Not editable pattern part of the text. Defaults to "".


## Usage

### Example

```swift

class SimpleController: UIViewController {
	
	@IBOutlet var urlField: BxTextField!
	@IBOutlet var phoneField: BxTextField!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        urlField.leftPatternText = "http://"
        urlField.rightPatternText = ".com"
        urlField.addTarget(self, action: #selector(urlDidChange(sender:)), for: .editingChanged)
        
        phoneField.leftPatternText = "+4 "
        phoneField.formattingReplacementChar = "*"
        phoneField.formattingTemplate = "(***) - ** - ** - **"
        phoneField.formattingEnteredCharacters = "0123456789"
        phoneField.addTarget(self, action: #selector(phoneDidChange(sender:)), for: .editingChanged)
    }
    
    @IBAction func urlDidChange (sender: BxTextField) {
        print(sender.enteredText) // it should show "your.inputed.domain.only"
    }
    
    @IBAction func phoneDidChange (sender: BxTextField) {
        print(sender.text) // it should show "+4 (123) - 45 - 67 - 89"
    }
    
}

```

### Different filling direction

```swift

class BxTextFieldFormattingTests: XCTestCase {

    func testRightToLeftDirection() {
        let textField = BxTextField(frame: CGRect(x: 0, y: 0, width: 200, height: 40))
        textField.formattingTemplate = "#.##"
        textField.formattingDirection = .rightToLeft
        textField.enteredText = "1"
        XCTAssertEqual(textField.text!, "1")
        textField.enteredText = "12"
        XCTAssertEqual(textField.text!, ".12")
        textField.enteredText = "123"
        XCTAssertEqual(textField.text!, "1.23")
    }
    
    func testLeftToRightDirection() {
        let textField = BxTextField(frame: CGRect(x: 0, y: 0, width: 200, height: 40))
        textField.formattingTemplate = "#.##"
        textField.formattingDirection = .leftToRight
        textField.enteredText = "1"
        XCTAssertEqual(textField.text!, "1")
        textField.enteredText = "12"
        XCTAssertEqual(textField.text!, "1.2")
        textField.enteredText = "123"
        XCTAssertEqual(textField.text!, "1.23")
    }
}

```

## License

BxTextField is released under the MIT license. See LICENSE for details.


CHANGELOG.md
# Copyright

BxTextField visual component for putting a text

ByteriX, 2017-2018. All right reserved.

# Versions

## 1.12.4 (10.06.2023)
##### Improvements
* support SPM

## 1.12 (08.06.2023)
##### Bug fix
* minimal target to iOS 11

## 1.11 (05.04.2021)
##### Improvements
* added activeLineWidth
* edited updating from changing
* memory leaks tests
* supporting iOS 14

## 1.10.12 (24.04.2020)
##### Improvements
* change description: supporting section
* add Swift 3.2 to supporting

## 1.10.11 (24.04.2020)
##### Bug fix
* edited PodSpec with supporting Swift versions section

## 1.10.10 (11.06.2019)
##### Improvements
* updated build.sh

## 1.10.9 (11.06.2019)
##### Improvements
* swift_versions changed

## 1.10.8 (07.06.2019)
##### Improvements
* rollback to old pods
* fixed build

## 1.10.4 (07.06.2019)
##### Bug fix
* fixed the placeholder shown

## 1.10.3 (05.06.2019)
##### Improvements
* fixed issue with title updating when text was full
* fixed issue with title animation when label was cuted
* added activationDuration as UIAppearance for changing animation

## 1.10.2 (16.05.2019)
##### Bug fix
* fixed a cutting of title text

## 1.10.1 (16.04.2019)
##### Improvements
* added isTitleUpper for the transformation title when it has went from hint state

## 1.10.0 (15.04.2019)
##### Improvements
* added decoration elements: line and title as is as Material designe
* more properties for settings control: title, titleAsHintFont, titleFont, titleSeporatorHeight, lineSeporatorHeight, activeColor, inactiveColor, lineColor, lineWidth
* isActive, isTitleAsHint information properties

## 1.9.0 (29.03.2019)
##### Bug fix
* fixed all warnings
* supporting Swift 5 / iOS 12.2 / XCode 10.2
* fixed unit tests

## 1.8.1 (12.11.2018)
##### Bug fix
* Updating placeholder when have changed placeholderColor

## 1.8.0 (04.10.2018)
##### Improvements
* Supporting Swift 4.2 and iOS 12
* Refactoring older version supporting, but left it

## 1.7.3 (11.09.2018)
##### Improvements
* Documentation with markdown notation
* covarage Unit/UI tests of formattingEnteredCharacters

## 1.7.2 (29.08.2018)
##### Bug fix
* Warning: All symboles how contains in formattingTemplate will be ignored from putting because it had broken alghoritm of putting

## 1.7.1 (13.08.2018)
##### Improvements
* formattingEnteredCharSet can be used if formattingTemplate is empty

## 1.7.0 (16.05.2018)
##### Improvements
* added isPlaceholderPatternShown property for magage pattern showing on placeholder
* test covarage of placeholder changing

## 1.6.2 (07.03.2018)
##### Bug fix
* new public method getEnteredText
* fixed enteredText property when we get unformatted text
* New test for getting nil text

## 1.6.1 (10.02.2018)
##### Bug fix
* bug fixed when changed enteredText from code
* made full general test covarage
* added extrem unit tests
* optimized getFormattedText function

## 1.6.0 (04.02.2018)
##### Improvements
* added isFormattingRewriting property with false default value for limitation formating putting
* added Unit Tests and UI Tests for testing isFormattingRewriting
* changed README

## 1.5.3 (29.01.2018)
##### Improvements
* added UI Test cases with leftToRight direction filing

## 1.5.2 (24.01.2018)
##### Improvements
* added UI Test cases with rightToLeft direction filing

## 1.5.1 (22.01.2018)
##### Improvements
* removed unused delegate class
* added Tests for String and Formatting components
* changed Readme, added example with filling direction

## 1.5.0 (09.01.2018)
##### Improvements
* fixed supporting Swift 3 and compiling/testing on XCode 8.x
* made String extension for supporting Swift 3/4
* added foundation tests
* clean all warnings
* refactoring and small description changes

## 1.4.11 (26.12.2017)
##### Improvements
* added secondary Tests

## 1.4.10 (25.12.2017)
##### Improvements
* added common Tests

## 1.4.9 (25.12.2017)
##### Bug fix
* fixed Tests target

## 1.4.8 (25.12.2017)
##### Improvements
* added Tests

## 1.4.7 (21.12.2017)
##### Improvements
* remade gif url absolutable for Readme

## 1.4.6 (21.12.2017)
##### Improvements
* test build script fix

## 1.4.5 (21.12.2017)
##### Improvements
* test old gif url
* changed build script

## 1.4.4 (14.12.2017)
##### Improvements
* changed common describtion Readme

## 1.4.3 (14.12.2017)
##### Bug fix
* remade gif url absolutable for Readme

## 1.4.2 (14.12.2017)
##### Bug fix
* made gif url absolutable for Readme

## 1.4.1 (08.12.2017)
##### Implement
* changed only description

## 1.4.0 (09.11.2017)
##### Implement
* Supporting Swift 4 with Swift 3
* Change formattingReplacementChar property to String
* Fixed some warnings

## 1.3.4 (19.09.2017)
##### Bug fix
* Fixed showing placeholder from iOS 11

## 1.3.3 (18.09.2017)
##### Bug fix
* Fixed updating placeholder after changing enteredTextColor/enteredTextFont/patternTextFont/patternTextColor

## 1.3.2 (14.09.2017)
##### Implement
* Swift 3.2 with Swift 3.1 supporting

## 1.3.1 (14.09.2017)
##### Implement
* backword Swift 3.1 supporting

## 1.3.0 (12.09.2017)
##### Implement
* Swift 3.2 supporting

## 1.2.5 (02.09.2017)
##### Bug fix
* fixed updating font and text. It was something bug from Apple

## 1.2.4 (22.06.2017)
##### Improvements
* added placeholderColor for a changing color of placeholder

## 1.2.3 (29.05.2017)
##### Improvements
* added formattingDirection for change direction of the putting
* refactored code
* bug with double assign rightPatternText fixed

## 1.1.4 (15.03.2017)
##### Bug fix
* fixed bug with showing the placeholder

## 1.1.3 (14.03.2017)
##### Implementing
* add edge property marginSize for showing border

## 1.1.2 (14.03.2017)
##### Refactoring
* change formated to formatted

## 1.1.1 (14.03.2017)
##### Bug fix
* updating content from properties font, textColor, text
* fixed optional attributes
* corrected logic of content showing

## 1.1.0 (06.03.2017)
##### Implementing
* refactoring: clear formatting function
* descripted all sources

BxTextFieldTests/BxTextFieldTests.swift
//
//  BxTextFieldTests.swift
//  BxTextFieldTests
//
//  Created by Sergey Balalaev on 02/03/17.
//  Copyright © 2017 Byterix. All rights reserved.
//

import XCTest
@testable import BxTextField

class BxTextFieldTests: XCTestCase {
    
    override func setUp() {
        super.setUp()
        // Put setup code here. This method is called before the invocation of each test method in the class.
    }
    
    override func tearDown() {
        // Put teardown code here. This method is called after the invocation of each test method in the class.
        super.tearDown()
    }
    
    func testRightPattern() {
        let textField = BxTextField(frame: CGRect(x: 0, y: 0, width: 200, height: 40))
        textField.rightPatternText = ".byterix.com"
        textField.enteredText = "mail"
        XCTAssertEqual(textField.text!, "mail.byterix.com")
    }
    
    func testLeftPattern() {
        let textField = BxTextField(frame: CGRect(x: 0, y: 0, width: 200, height: 40))
        textField.leftPatternText = "www."
        textField.enteredText = "byterix.com"
        XCTAssertEqual(textField.text!, "www.byterix.com")
    }
    
    func testAllPattern() {
        let textField = BxTextField(frame: CGRect(x: 0, y: 0, width: 200, height: 40))
        textField.leftPatternText = "http://"
        textField.rightPatternText = "byterix.com"
        textField.enteredText = ""
        XCTAssertEqual(textField.text!, "http://byterix.com")
    }
    
    func testFormatting() {
        let textField = BxTextField(frame: CGRect(x: 0, y: 0, width: 200, height: 40))
        textField.formattingTemplate = "+#(###)###-##-##"
        textField.enteredText = "78901234567"
        XCTAssertEqual(textField.text!, "+7(890)123-45-67")
    }
    
    func testNilText() {
        let textField = BxTextField(frame: CGRect(x: 0, y: 0, width: 200, height: 40))
        textField.formattingTemplate = "+#(###)###-##-##"
        textField.text = nil
        XCTAssertEqual(textField.text!, "")
        XCTAssertEqual(textField.enteredText, "")
        textField.enteredText = "123"
        XCTAssertEqual(textField.text!, "+1(23")
        textField.text = nil
        XCTAssertEqual(textField.text!, "")
        XCTAssertEqual(textField.enteredText, "")
    }
    
    func testBackwardTextPutting() {
        let textField = BxTextField(frame: CGRect(x: 0, y: 0, width: 200, height: 40))
        textField.formattingTemplate = "+#(###)###-##-##"
        textField.formattingEnteredCharacters="0123456789"
        textField.enteredText = "12345"
        XCTAssertEqual(textField.text!, "+1(234)5")
        XCTAssertEqual(textField.enteredText, "12345")
        textField.text = "+7(654)3"
        XCTAssertEqual(textField.text!, "+7(654)3")
        XCTAssertEqual(textField.enteredText, "76543")
    }
    
    func testNotEnteredTextPutting() {
        let textField = BxTextField(frame: CGRect(x: 0, y: 0, width: 200, height: 40))
        textField.formattingTemplate = "+#(###)###-##-##"
        textField.formattingEnteredCharacters="0123456789"
        textField.enteredText = "(abc1-+-W2R3S4###())5"
        XCTAssertEqual(textField.text!, "+1(234)5")
        XCTAssertEqual(textField.enteredText, "12345")
        textField.text = "+7#(654)3W"
        XCTAssertEqual(textField.text!, "+7(654)3")
        XCTAssertEqual(textField.enteredText, "76543")
    }
    
    func testNotEnteredTextPuttingFromTemplate() {
        let textField = BxTextField(frame: CGRect(x: 0, y: 0, width: 200, height: 40))
        textField.formattingTemplate = "+#(###)###-##-##"
        textField.enteredText = "()-#123-()#45-#()"
        XCTAssertEqual(textField.text!, "+1(234)5")
        XCTAssertEqual(textField.enteredText, "12345")
    }
    
    func testAddToEnteredTextTemplateSymboles() {
        let textField = BxTextField(frame: CGRect(x: 0, y: 0, width: 200, height: 40))
        textField.formattingTemplate = "+#(###)###-##-##"
        textField.formattingEnteredCharacters="0123456789+-#()"
        textField.enteredText = "()-#123-()#45-#()"
        XCTAssertEqual(textField.text!, "+1(234)5")
        XCTAssertEqual(textField.enteredText, "12345")
    }

    
}
